single_cell_RNAseq_snake/bin/de_pool.R

```diff
@@ -0,0 +1,89 @@
+# This script depools demultiplexed single cell RNA seq data
+
+load_seurat_objects <- function(file_paths) {
+    seurat_objects <- list()
+
+    for (file_path in file_paths) {
+        if (file.exists(file_path)) {
+            seurat_objects[[basename(file_path)]] <- readRDS(file_path)
+        } else {
+            stop("File not found: ", file_path)
+        }
+    }
+    return(seurat_objects)
+}
+
+
+save_seurat_objects <- function(filename, seurat_list) {
+  # Check if the provided list is actually a list of Seurat objects
+  if (!is.list(seurat_list)) {
+    stop("The provided 'seurat_list' is not a list.")
+  }
+
+  for (obj in seurat_list) {
+    if (!inherits(obj, "Seurat")) {
+      stop("All elements of 'seurat_list' must be Seurat objects.")
+    }
+  }
+
+  dir_to_create <- dirname(filename) # Extract the directory part of the filename
+
+  # Create the directory if it doesn't exist
+  if (!dir.exists(dir_to_create)) {
+      dir.create(dir_to_create, recursive = TRUE)
+  }
+
+  # Save the list of Seurat objects to a file
+  saveRDS(seurat_list, file = filename)
+}
+
+
+depool_by_id <- function(seurat_objects) {
+  depooled_objs <- list()
+  
+  for (seurat_obj in seurat_objects) {
+    # Extract metadata
+    metadata <- seurat_obj@meta.data
+    
+    # Split BEST.GUESS values and extract first for SNG, then store in 'sample_id'
+    metadata$sample_id <- sapply(strsplit(as.character(metadata$BEST.GUESS), ","), function(x) x[1])
+    seurat_obj@meta.data <- metadata  # Update the metadata in the Seurat object
```

Comment:
be careful with updating the Seurat metadata -- you can lose the rownames which are the cell ids which creates problems downstream.
Instead I'd use `Seurat::AddMetaData()`

Reply:
good catch ! Thank you